### Binary Exponentiation

- Imagine we are asked to compute $a^n$

- Of course, in the usual brute force approach, we need to run a loop n times

In [5]:
def binary_exponentiation_naive(base: int, exponent: int) -> int:
    '''
    Time complexity: O(N), where N is the value of the exponent
    '''
    retval = 1
    for i in range(exponent):
        retval *= base
    return retval

binary_exponentiation_naive(2, 13)

8192

- Notice that most of the time comes from the loop! So we want to rewrite the exponentiation such that the loop is minimised. 
    - Imagine we want to do exponentiation of $2^{13}$. Rules:
        - Whenever exponent is odd, multiply the `result` by the base, and subtract one from the exponent 
        - Whenever exponent is even, square the base, and halve the power 
        - Keep going until power reaches 0 

    - Example: Recall that 2^13 previously took 13 steps (because it loops 13 times). Now, we simply take 7 steps!  
        - result = 1; base = 2; power = 13
        - result = 2; base = 2; power = 12
        - result = 2; base = 4; power = 6
        - result = 2; base = 16; power = 3
        - result = 32; base = 16; power = 2
        - result = 32; base = 256; power = 1
        - result = 8192; base = 256; power = 0

    - Time complexity:
        - Let the exponent be $N$
        - We reduce the exponent by half every step, so we have complexity of $O(\log_2(N))$

In [4]:
def binary_exponentiation(base: int, exponent: int) -> int:
    '''
    Time complexity: O(log(N)) due to halving of exponent every step
    '''
    result = 1

    while exponent > 0:
        if exponent % 2 == 1:
            result *= base
        
        exponent = exponent // 2
        base *= base
    
    return result

binary_exponentiation(2, 13)

8192

### Extension: Modular Exponentiation

- A simple extension of this exponentiation is when we wish to take the mod of the exponentiation; $A^B \mod C$
    - Recall from modular arithmetic that when $A \equiv B (\mod N)$, then $A^k \equiv B^k (\mod N)$
    - It is trivially true that $A \equiv A (\mod N)$
    - Then 
